### PR TITLE
Ergonomic format_args!

### LINKED ISSUE #9456: format! should support multiple formats for one argument
Take something like the following format string:

```rust
format!("{a:?} is {a}", a = "foo\u2014bar")
```

I *expect* this to produce the string `"foo\u2014bar" is foo—bar`.

However, what it actually produces is `"foo\u2014bar" is "foo\u2014bar"`; the behaviour of the second placeholder has changed from using `Default` to using `Poly`.

If I make the type difference explicit with the format string `"{a:?} is {a:s}"`, then I get a compilation error, which, although not what I wanted, at least *tells* me I can't do what I'm trying to do:

```
0.rs:2:50: 2:64 error: argument redeclared with type `s` when it was previously `?`
0.rs:2     println!("a is {a:?} (literally, {a:s})", a = "foo\u2014bar");
                                                         ^~~~~~~~~~~~~~
error: aborting due to previous error
```

My problem is with the implicit change of behaviour for named arguments from using `Default` to using what was done last. This distinctly confused me when I came across it. If using an argument with different formatting constraints is not possible (I don't see why this constraint is there, truth to tell) then I believe the unspecified format should still be interpreted as `Default` rather than what-was-done-last, and an error raised rather than this surprising behaviour.

### PR DESCRIPTION
**Pull request by xen0n**
Fixes #9456 (at last).

Not a ground-up rewrite of the existing machinery, but more like an added intermediary layer between macro arguments and format placeholders. This is now implementing Rust RFC 1618!

### CODE REVIEW

**Comment by xen0n** [comment]
This caused some small regressions here and there. I'll be slowly fixing them; compilation time is very long.

Non-regressions, just TODO:

* [ ] `compile-fail/ifmt-bad-arg.rs`
* [ ] `run-pass/ifmt.rs`

Noticed regressions:

* [ ] `compile-fail/macro-backtrace-println.rs`
* [ ] doctest `fmt_3` (`{:.*}` didn't eat the precision argument)

**Comment by xen0n** [comment]
Implemented correct implicit args handling by resolving all `CountIsParam(i)`s and `ArgumentNext`s into absolute references during parse. Which means, for example, `{foo:.*} {} {1:.*} {:.*} {bar:.*}` now parses exactly like `{foo:.0$} {1} {1:.2$} {4:.3$} {bar:.5$}`. Which obviates the need to handle any next-references during expansion and outputting; will remove these later, after I finish up an RFC for this PR later this week.

**Comment by alexcrichton** [comment]
Thanks for the PR! I'll take a closer look at this once https://github.com/rust-lang/rfcs/pull/1618 is handled

**Comment by xen0n** [comment]
Fixed test cases broken by this PR, and added new cases as well.

Currently `make check` fails on several unrelated tests:

* `run-pass/issue-28950.rs`
* `run-pass/panic-runtime/lto-abort.rs`
* `run-pass/panic-runtime/lto-unwind.rs`
* `run-pass/sepcomp-lib-lto.rs`

I'm not sure why those keep popping up, it wasn't the case several days ago. Testing master branch right now, will edit this once finished.

**Comment by xen0n** [comment]
Travis test completed without the unrelated failures, seems there is a pretty-print round-trip failure:

```diff
--- pretty.expected.rs	2016-05-22 13:39:14.080831211 +0800
+++ pretty.actual.rs	2016-05-22 13:38:56.912831929 +0800
@@ -178,7 +178,6 @@
        , b = "hijklmn" , c = 3 ) , "abcd hijk 4\nabc hij 3");
     t!(format ! ( "{a:.*} {0} {:.*}" , 4 , 3 , "efgh" , a = "abcdef" ) ,
        "abcd 4 efg");
-
     // Test that pointers don't get truncated.
     {
         let val = usize::MAX;
```

Apart from this and deduplication of argument objects, the feature feels almost ready. Waiting for RFC resolution :smile:

**Comment by alexcrichton** [comment]
Ah yeah unfortunately we've had bad luck with debuginfo and LTO in the past (lots of weird LLVM assertions), but thanks for the progress update!

**Comment by xen0n** [comment]
I immediately gave up fixing the pretty print issue after trying to produce a minimal test case and got this:

```rust
macro_rules! t {
    ($a:expr, $b:expr) => { assert_eq!($a, $b) }
}

pub fn main() {
    t!(format!("{:x}{:X}{:x}{:X}{:x}{a:X}",1,2,3,4,5,a=15), "abcdefghijkl");




    // every time this code get pretty-printed, a newline is eaten!
}
```

Might as well reduce the length of affected lines, as the bug seems to only occur when a line wrap is needed in the output. I'll finish the argument object deduplication and then we can start reviewing.

**Comment by xen0n** [comment]
Rebased and cleaned up a little, should be ready for review!


PS: The reason I think an RFC is really needed, is that some serious documentation changes and maybe deprecation inside the formatting runtime will have to be done. However as the RFC is already submitted and waiting through its FCP this point is already moot.

File: src/test/run-pass/ifmt.rs [hunk 1]
```diff
@@ -163,6 +163,28 @@ pub fn main() {
...
+               4, a="abcdefg", b="hijklmn", c=3),
+               "abcd hijk 4\nabc hij 3");
+    t!(format!("{a:.*} {0} {:.*}", 4, 3, "efgh", a="abcdef"), "abcd 4 efg");
+    t!(format!("{:.a$} {a} {a:#x}", "aaaaaa", a=2), "aa 2 0x2");
+
+    // And don't ICE for the following case
+    // named arg only used as count
+    t!(format!("{0:.a$}", "aaaaaa", a=2), "aa");
+
+
```
Comment (alexcrichton) [inline src/test/run-pass/ifmt.rs]: Can you add a test here as well with side effects on the evaluated expression to ensure it's only calculated once even though it may be formatted multiple times.

File: src/libsyntax_ext/format.rs [hunk 2]
```diff
@@ -626,7 +753,9 @@ pub fn expand_preparsed_format_args(ecx: &mut ExtCtxt, sp: Span,
                                     name_ordering: Vec<String>,
                                     names: HashMap<String, P<ast::Expr>>)
                                     -> P<ast::Expr> {
-    let arg_types: Vec<_> = (0..args.len()).map(|_| None).collect();
+    // We can't use vec! here for Vec<ArgumentType>, as ArgumentType isn't Clone.
+    let arg_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
+    let arg_unique_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: I believe this can just be `vec![Vec::new(); args.len()]`

File: src/libfmt_macros/lib.rs [hunk 3]
```diff
@@ -77,9 +77,39 @@ pub struct FormatSpec<'a> {
     pub ty: &'a str,
 }
 
+/// For internal use; same as FormatSpec except using CountInternal instead
+/// of Count. Possible next-references in the count fields will be resolved
+/// before the struct is converted into FormatSpec for returning.
+struct FormatSpecInternal<'a> {
```
Comment (alexcrichton) [inline src/libfmt_macros/lib.rs]: Could we perhaps not add this and just handle all conversions and such in-place during parsing?

File: src/libsyntax_ext/format.rs [hunk 4]
```diff
@@ -46,13 +47,14 @@ struct Context<'a, 'b:'a> {
     /// Parsed argument expressions and the types that we've found so far for
     /// them.
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Can you update this documentation for what `arg_types` and `arg_unique_types` are?

File: src/libsyntax_ext/format.rs [hunk 5]
```diff
@@ -83,9 +98,9 @@ struct Context<'a, 'b:'a> {
 fn parse_args(ecx: &mut ExtCtxt, sp: Span, tts: &[ast::TokenTree])
               -> Option<(P<ast::Expr>, Vec<P<ast::Expr>>, Vec<String>,
                          HashMap<String, P<ast::Expr>>)> {
-    let mut args = Vec::new();
+    let mut args = vec![];
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Could changes like this be left out?

File: src/libfmt_macros/lib.rs (hunk 3 above)
Comment (xen0n) [inline src/libfmt_macros/lib.rs]: We don't have private enum variants at the moment, and I don't want to add a lot of `_` match arms to other places where the enums are consumed. But I'll look into this later though, currently outside.

**Comment by alexcrichton** [comment]
This seems to have gotten pretty significantly more complicated with some unfortunate code duplication in a few places, would it be possible perhaps reduce the number of maps and index pointers in play? One easy optimization would be to just translate all named arguments immediately to numbered indexes, that way there's no branch for each.

File: src/libsyntax_ext/format.rs (hunk 2 above)
Comment (xen0n) [inline src/libsyntax_ext/format.rs]: Nope, `ArgumentType` is not `Clone`, so `vec![something; num]`, which is `vec::from_elem` under the hood, will fail on this.

```rust
#[derive(PartialEq)]
enum ArgumentType {
    Placeholder(String),
    Count,
}

fn main() {
    let a: Vec<Vec<ArgumentType>> = vec![Vec::new(); 100];
}
```

No matter how you say `Vec::new()` this won't compile. Either we make `ArgumentType` `Clone` which I assume is not really desirable, or keep using the current way, which isn't that ugly after all.

**Comment by xen0n** [comment]
> One easy optimization would be to just translate all named arguments immediately to numbered indexes, that way there's no branch for each. 

Nice insight! Will try to decrease the complexity this way. Previously I wanted to keep the original machinery intact as much as possible, but the patch have grown beyond my expectation.

**Comment by xen0n** [comment]
Addressed the review comments. By applying the named argument optimization I was able to trim down the patch significantly (280 -> 113 net added lines)! I'll squash the review fix commits once everything looks good.

File: src/libfmt_macros/lib.rs [hunk 6]
```diff
@@ -259,9 +262,41 @@ impl<'a> Parser<'a> {
...
+        // incompatibilities.
+        match format.precision {
+            CountIsNextParam => {
+                // eat the current implicit arg
+                let i = self.curarg;
+                self.curarg += 1;
+                format.precision = CountIsParam(i);
+            }
+            _ => {}
+        }
```
Comment (alexcrichton) [inline src/libfmt_macros/lib.rs]: I wonder if it might be best to just remove the `CountIsNextParam` and `ArgumentNext` enum variants? That way this could be handled directly in the `position` and `format` methods

File: src/libsyntax_ext/format.rs [hunk 7]
```diff
@@ -131,24 +145,49 @@ fn parse_args(ecx: &mut ExtCtxt, sp: Span, tts: &[ast::TokenTree])
...
-    Some((fmtstr, args, order, names))
+    Some((fmtstr, args, names))
 }
 
 impl<'a, 'b> Context<'a, 'b> {
-    /// Verifies one piece of a parse string. All errors are not emitted as
-    /// fatal so we can continue giving errors about this and possibly other
-    /// format strings.
+    fn resolve_name_inplace(&self, p: &mut parse::Piece) {
+        let lookup = |s| *self.names.get(s).unwrap_or(&0);
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Is this `unwrap_or` branch ever taken? It seems like this may want to be `unwrap()` to indicate a bug

**Comment by alexcrichton** [comment]
Thanks @xen0n! I'll take a closer look once the RFC is resolved.

File: src/libsyntax_ext/format.rs (hunk 7 above)
Comment (xen0n) [inline src/libsyntax_ext/format.rs]: It's taken during parse of illegal invocations like in `format_args!("{foo}")` where the piece tries to reference a non-existent named argument, so this is necessary. Maybe I should add a comment regarding this.

File: src/libfmt_macros/lib.rs (hunk 6 above)
Comment (xen0n) [inline src/libfmt_macros/lib.rs]: I'm afraid we can't, as references to them are already indirectly pulled into existing compiled libraries, thus removing them would mean an ABI breakage. The best we could do is to not use them in newer code and hope that it'll die out some time.

File: src/libsyntax_ext/format.rs (hunk 7 above)
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Ah that makes sense, and yeah feel free to add a comment!

File: src/libfmt_macros/lib.rs (hunk 6 above)
Comment (alexcrichton) [inline src/libfmt_macros/lib.rs]: Oh the `fmt_macros` library is actually entirely unstable so we can change the APIs whenever we want, no need to preserve backwards compatibility

Comment (xen0n) [inline src/libfmt_macros/lib.rs]: So Rust doesn't allow intermixing object code from different compiler versions? I'll try to do that without private enum variants though.

Comment (alexcrichton) [inline src/libfmt_macros/lib.rs]: Yeah we don't need to worry about sharing like that. It should be fine to just drop these enum variants in `fmt_macros`

File: src/libsyntax_ext/format.rs [hunk 8]
```diff
@@ -211,84 +244,70 @@ impl<'a, 'b> Context<'a, 'b> {
...
+                                seen_ty.push(ty);
+                                i
+                            }
+                        };
+                        self.arg_types[arg].push(i);
+                    }
+                    Count => {
+                        match self.count_positions.entry(arg) {
+                            Entry::Vacant(e) => {
+                                let i = self.count_positions_count;
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Does this mean that `count_positions_count` an alias for the length of the `count_positions` map?

File: src/libsyntax_ext/format.rs [hunk 9]
```diff
@@ -211,84 +244,70 @@ impl<'a, 'b> Context<'a, 'b> {
...
+        let args_len = self.args.len();
+        self.arg_index_map.reserve(args_len);
+
+        let mut sofar = 0usize;
+
+        // Map the arguments
+        for i in 0..args_len {
+            let ref arg_types = self.arg_types[i];
+            let mut arg_offsets = Vec::with_capacity(arg_types.len());
+            for ref offset in arg_types {
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: I think this `ref` can be elided as `arg_types` is already `&Vec<T>` so it yields references when iterated over.

File: src/libsyntax_ext/format.rs [hunk 10]
```diff
@@ -357,25 +380,34 @@ impl<'a, 'b> Context<'a, 'b> {
...
-                        // at the end of the list of arguments
-                        parse::ArgumentNamed(n) => {
-                            let i = match self.name_positions.get(n) {
-                                Some(&i) => i,
-                                None => 0, // error already emitted elsewhere
+                        parse::ArgumentNext => panic!("should never happen"),
+                        parse::ArgumentIs(i) => {
+                            // Map to index in final generated argument array
+                            // in case of multiple types specified
+                            let arg_idx = if self.args.len() > i {
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: This `if` may be best expressed with some form of `<[T]>::get` plus a `match`

File: src/libsyntax_ext/format.rs [hunk 11]
```diff
@@ -623,10 +650,10 @@ pub fn expand_format_args<'cx>(ecx: &'cx mut ExtCtxt, sp: Span,
 pub fn expand_preparsed_format_args(ecx: &mut ExtCtxt, sp: Span,
                                     efmt: P<ast::Expr>,
                                     args: Vec<P<ast::Expr>>,
-                                    name_ordering: Vec<String>,
-                                    names: HashMap<String, P<ast::Expr>>)
+                                    names: HashMap<String, usize>)
                                     -> P<ast::Expr> {
-    let arg_types: Vec<_> = (0..args.len()).map(|_| None).collect();
+    let arg_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
+    let arg_unique_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Nowadays this may be best expressed as:

```rust
vec![Vec::new(); args.len()]
```

File: src/libsyntax_ext/format.rs [hunk 12]
```diff
@@ -43,17 +44,16 @@ struct Context<'a, 'b:'a> {
...
     args: Vec<P<ast::Expr>>,
-    arg_types: Vec<Option<ArgumentType>>,
-    /// Parsed named expressions and the types that we've found for them so far.
-    /// Note that we keep a side-array of the ordering of the named arguments
-    /// found to be sure that we can translate them in the same order that they
-    /// were declared in.
-    names: HashMap<String, P<ast::Expr>>,
-    name_types: HashMap<String, ArgumentType>,
-    name_ordering: Vec<String>,
+    /// Placeholder slot numbers indexed by argument.
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Could you expand this a bit as well? I found this kinda hard to follow, but it looks like if you have a format string like `{0:x} {0:X} {}` this would be something like:

```
[ 
    [0, 1],
    [0],
]
```

where the indexes are pointers into `arg_unique_types`?

File: src/libsyntax_ext/format.rs [hunk 13]
```diff
@@ -43,17 +44,16 @@ struct Context<'a, 'b:'a> {
...
+    arg_types: Vec<Vec<usize>>,
+    /// Unique format specs seen for each argument.
+    arg_unique_types: Vec<Vec<ArgumentType>>,
+    /// Map from named arguments to their resolved indices.
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: There's quite a few indexes flying around here, so it may be worth expanding on exactly what "indice" this is containing?

File: src/libsyntax_ext/format.rs [hunk 14]
```diff
@@ -65,10 +65,27 @@ struct Context<'a, 'b:'a> {
...
+
+    /// Starting offset of count argument slots.
+    count_args_index_offset: usize,
+
+    /// Count argument slots and tracking data structures.
+    count_args: Vec<Position>,
+    /// Relative slot numbers for count arguments.
+    count_positions: HashMap<usize, usize>,
+    /// Number of count slots assigned.
+    count_positions_count: usize,
```
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: I may be getting confused, but before there were no entries in this structure tracking `count_*` fields. How come these all cropped up now?

**Comment by xen0n** [comment]
@alexcrichton Just a quick note that I might need another 2~3 days in completing this; I have another project going at the moment (just done a CyanogenMod port few days ago and already tester feedbacks are pouring at me). I'll ping the issue with commits when I'm done!

File: src/libsyntax_ext/format.rs (hunk 11 above)
Comment (xen0n) [inline src/libsyntax_ext/format.rs]: That would require `ArgumentType` to also derive `Clone`. Is this acceptable, as we actually never clone instances of it?

**Comment by xen0n** [comment]
Rebased and addressed most of review comments pending:

* [ ] decision whether `ArgumentType` needs to derive `Clone` only for saner initialization of `Vec<Vec<ArgumentType>>`
* [ ] comment and documentation fixes

File: src/libsyntax_ext/format.rs (hunk 11 above)
Comment (alexcrichton) [inline src/libsyntax_ext/format.rs]: Ah right, that's fine to leave this as-is then, no need to add `Clone`

**Comment by xen0n** [comment]
Rebased again and (finally found time to do so!) added comments to facilitate better understanding. Hope this is not too late for 1.10!

**Comment by alexcrichton** [comment]
Oops, sorry this fell off the radar! I think this got lost in my inbox by accident...

Ok, everything looks good to me, thanks @xen0n! Could you also squash the commits down? Other than that r=me

**Comment by xen0n** [comment]
Squashed all review-generated commits into respective commits and rebased again. Also clarified a few `_ => panic!` cases for better maintainability (no worries if some enum variants got added someday). Should be good to go now! @alexcrichton

**Comment by alexcrichton** [comment]
@bors: r+ 51e54e57a45d7b956e3e3390db76ea2651f5d3b3

Thanks again!
